File: 62.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        def find(i: int, j: int, m: int, n: int, dp: List[List[int]]) -> int:
            if i == m - 1 and j == n - 1:
                return 1
            if i == m or j == n:
                return 0
            if dp[i][j] != 0:
                return dp[i][j]
            dp[i][j] = find(i + 1, j, m, n, dp) + find(i, j + 1, m, n, dp)
            return dp[i][j]

        dp = [[]] * m
        for i in range(m):
            dp[i] = [0] * n
        return find(0, 0, m, n, dp)

File: java/62.java
class Solution {
    public int uniquePaths(int m, int n) {
        return find(0, 0, m, n, new int[m][n]);
    }

    private int find(int i, int j, int m, int n, int dp[][]) {
        if (i == m - 1 && j == n - 1)
            return 1;
        if (i == m || j == n)
            return 0;
        if (dp[i][j] != 0)
            return dp[i][j];
        return dp[i][j] = find(i + 1, j, m, n, dp) + find(i, j + 1, m, n, dp);
    }
}
